### PR TITLE
feat: typescript typings and example

### PR DESCRIPTION
**Pull request by NickBolles**

### CODE REVIEW

**Comment by NickBolles** [comment]
Followup - Remove package from DefinitelyTyped - https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package

**Comment by NickBolles** [comment]
reminder to look at changes to the @types package #40778 linked above and https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40840 before merging

**Comment by davidlandais** [comment]
Why are you waiting to merge this PR ? 👍

File: package.json [hunk 1]
```diff
@@ -25,6 +26,8 @@
     "heroku-postbuild": "nuxt build examples/demo",
     "start": "nuxt start examples/demo",
     "dev": "nuxt examples/demo",
+    "postinstall": "cd examples/with-typescript && npm install",
```
Comment (pi0) [inline package.json]: I think this is not required and also we use yarn :)

File: package.json [hunk 2]
```diff
@@ -71,7 +75,7 @@
     "jest": "^24.9.0",
     "jsdom": "^15.1.1",
     "nuxt-class-component": "^1.3.0",
-    "nuxt-edge": "^2.10.0-26144655.e9c4bcfe",
+    "nuxt-edge": "^2.10.3-26236810.6cc962fa",
```
Comment (pi0) [inline package.json]: Dependency upgrades are handled by renovate bot

File: docs/guide/setup.md [hunk 3]
```diff
@@ -28,3 +28,24 @@ auth: {
 ::: warning IMPORTANT
 When adding `auth-module` to a new Nuxt project ensure you have activated the Vuex store. More information on how to do that can be found on the [Nuxt Getting Started Guide](https://nuxtjs.org/guide/vuex-store).
 :::
+
+
+### Typescript setup
+
+Add the types to your "types" array in tsconfig.json after the `@nuxt/vue-app` entry
```
Comment (kevinmarrec) [inline docs/guide/setup.md]: Since Nuxt 2.9 it's `@nuxt/types` (https://typescript.nuxtjs.org/migration.html)
```suggestion
Add the types to your "types" array in tsconfig.json after the `@nuxt/types` entry
```

File: docs/guide/setup.md [hunk 4]
```diff
@@ -28,3 +28,24 @@ auth: {
...
+
+**tsconfig.json**
+
+```json
+{
+  "compilerOptions": {
+    "types": [
+      "@nuxt/vue-app",
```
Comment (kevinmarrec) [inline docs/guide/setup.md]:
```suggestion
      "@nuxt/types",
```

File: examples/with-typescript/mixins/auth.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+   * Logout method using auth-module with custom post-request
+   * logic, using toast module to show information, success
+   * and error messages.
+   *
+   * @returns {Promise<void>}
+   */
+  async logout () {
+
+    this.$toast.show('Logging out...')
+    await this.$auth.logout().then(() => {
```
Comment (kevinmarrec) [inline examples/with-typescript/mixins/auth.ts]: Doesn't seems logic to me to combine `await` with `then/catch`.
`$toast.success` shoud be called inside a `try` statement after `logout`, and `$toast.error` in the `catch` :
```ts
try {
  await this.$auth.logout()
  this.$toast.success('Successfully disconnected')
} catch (err) {
  this.$toast.error('Error while disconnecting: ' + err.message)
}

File: examples/with-typescript/mixins/auth.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+   * Login method using auth-module with custom post-request
+   * logic, using toast module to show information, success
+   * and error messages.
+   *
+   * @returns {Promise<T>}
+   */
+  async login() {
+    this.$toast.show('Log in...')
+    await this.$auth.login({
```
Comment (kevinmarrec) [inline examples/with-typescript/mixins/auth.ts]: Doesn't seems logic to me to combine `await` with `then/catch`.
`$toast.success` shoud be called inside a `try` statement after `login`, and `$toast.error` in the `catch` (also please notice the typo fix `connecting` instead of `disconnecting`) :
```ts
try {
  await this.$auth.login({
    data: {
      username: this.username,
      password: this.password
    }
  })
  this.$toast.success('Successfully connected')
} catch (err) {
  this.$toast.error('Error while connecting: ' + err.message)
  this.loginHasError = true
}

File: examples/with-typescript/nuxt.config.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      }
+    }
+  },
+  buildModules: ["@nuxt/typescript-build"],
+  modules: [
+    "bootstrap-vue/nuxt", // https://bootstrap-vue.js.org/docs/#nuxt-js
+    "@nuxtjs/axios",
+    "@nuxtjs/auth",
+    "@nuxtjs/toast",
+    "@@"
```
Comment (kevinmarrec) [inline examples/with-typescript/nuxt.config.ts]: I don't remember what `@@` is for 🤷‍♂️

File: examples/with-typescript/package.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  "main": "index.js",
+  "scripts": {
+    "dev": "nuxt-ts",
+    "build": "nuxt-ts build",
+    "generate": "nuxt-ts generate",
+    "start": "nuxt-ts start"
+  },
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
```
Comment (kevinmarrec) [inline examples/with-typescript/package.json]: devDependencies need to be upgraded
image

File: examples/with-typescript/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
```
Comment (kevinmarrec) [inline examples/with-typescript/tsconfig.json]: Target should be now `es2018` (or `es2019`/`es2020`) to support last TypeScript features, see note in [**Nuxt TypeScript documentation**](https://typescript.nuxtjs.org/guide/setup.html#configuration).

```suggestion
    "target": "es2018",
```

File: package.json (hunk 1 above)
Comment (kevinmarrec) [reply package.json]: I think that too

File: types/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  syncToken(strategyName: string): string;
+  onError(handler: (error: Error, name: string, endpoint: any) => void): any;
+  setUser(user?: Partial<T>): any;
+  reset(): Promise<never>;
+  redirect(name: string): any;
+}
+
+declare module "@nuxt/types" {
+  interface Context {
+    $axios: NuxtAxiosInstance;
```
Comment (kevinmarrec) [inline types/index.d.ts]: I think `NuxtAxiosInstance` needs be imported to be used here

File: examples/with-typescript/nuxt.config.ts (hunk 7 above)
Comment (pi0) [reply examples/with-typescript/nuxt.config.ts]: It is rootDir.

Comment (NickBolles) [reply examples/with-typescript/nuxt.config.ts]: So no changes here right? I pulled this from the other example I believe

File: examples/with-typescript/mixins/auth.ts (hunk 5 above)
Comment (NickBolles) [reply examples/with-typescript/mixins/auth.ts]: Agreed, this is copy and pasted from https://github.com/nuxt-community/auth-module/blob/dev/examples/with-class-component/mixins/auth.js, so I'll fix both of them

File: examples/with-typescript/mixins/auth.ts (hunk 6 above)
Comment (NickBolles) [reply examples/with-typescript/mixins/auth.ts]: See above comment

File: examples/with-typescript/package.json (hunk 8 above)
Comment (NickBolles) [reply examples/with-typescript/package.json]: Should be fixed by a rebase

File: examples/with-typescript/tsconfig.json (hunk 9 above)
Comment (NickBolles) [reply examples/with-typescript/tsconfig.json]: Sounds good.

File: package.json (hunk 1 above)
Comment (NickBolles) [reply package.json]: yea, this is bad... oops. Maybe I'll move yarn install to dev:ts and update dev:ts to use yarn to run too

**Comment by NickBolles** [comment]
As part of the conversion to use async as requested. I removed the comments suggesting async

```

    // If you are not fond of using axios promises on async calls
    // You can still use Javascript try and catch block
    //
    // try {
    //   this.$toast.show('Logging out...')
    //   await this.$auth.logout()
    //   this.$toast.success('Successfully disconnected')
    // } catch (err) {
    //   this.$toast.error('Error while connecting: ' + err.message)
    // }
```

File: types/index.d.ts (hunk 10 above)
Comment (NickBolles) [reply types/index.d.ts]: this was actually supposed to be `$auth: Auth`... oops.

**Comment by NickBolles** [comment]
@kevinmarrec @pi0

> Thanks for the review @kevinmarrec and @pi0 I'll look at these today, rebase and resubmit.

So much for "today"... All of these should be fixed. The examples should be working correctly too. Let me know if you see anything else!

**Comment by HappyStore** [comment]
Any updates?

File: lib/core/middleware.js [hunk 11]
```diff
@@ -16,7 +16,8 @@ Middleware.auth = function (ctx) {
 
   const { login, callback } = ctx.$auth.options.redirect
   const pageIsInGuestMode = routeOption(ctx.route, 'auth', 'guest')
-  const insidePage = page => normalizePath(ctx.route.path) === normalizePath(page)
+  const insidePage = page =>
```
Comment (pi0) [inline lib/core/middleware.js]: Is this change necessary?

File: lib/core/hook.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import { routeOption, getMatchedComponents, normalizePath } from './utilities'
```
Comment (pi0) [inline lib/core/hook.js]: - Cloning a Nuxt template into the module is not probably the best idea
- I would suggest creating a new issue first discussing on hook possibility and sperate from this PR
- A full rewrite is planned of auth module to use hook functions like this :)

Comment (NickBolles) [reply lib/core/hook.js]: Now that I'm looking back at this, I think this PR has changed from a really old change I was gonna propose, with this hook. Which sucks cuz I spent the time building this change off of that :(

This PR will require a rework then. I think the same concepts will apply, but I have to reset to master before doing them

Comment (NickBolles) [reply lib/core/hook.js]: These should be resolved now @pi0

**Comment by NickBolles** [comment]
@pi0 @kevinmarrec Could you guys take another look? somehow I merged two branches and added a bunch of incorrect changes in the middle of this PR. (probably cuz I'm on a dev branch and not a feature branch....) Should be good to go. 

Should also make changes like this: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43380 easier

File: lib/core/middleware.js (hunk 11 above)
Comment (NickBolles) [reply lib/core/middleware.js:19]: Removed, part of a different branch incorrectly included.

File: lib/core/hook.js (hunk 12 above)
Comment (NickBolles) [reply lib/core/hook.js]: Removed, part of a different branch incorrectly included.

File: types/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  $storage: Storage;
+  user: Partial<T>;
+  loggedIn: boolean;
+  loginWith(strategyName: string, ...args: any): Promise<never>;
+  login(...args: any): Promise<never>;
+  logout(): Promise<never>;
+  fetchUser(): Promise<never>;
+  fetchUserOnce(): Promise<never>;
+  hasScope(scopeName: string): boolean;
+  setToken(strategyName: string, token?: string): string;
```
Comment (JoaoPedroAS51) [inline types/index.d.ts:33]: `setToken(strategyName, token)`, `getToken(strategyName)` and `syncToken(strategyName)` has been deprecated. They were moved to Token class and refactored. You can use them like this `$auth.token.set(token)`, `$auth.token.get()` and `$auth.token.sync()`.

File: types/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  onError(handler: (error: Error, name: string, endpoint: any) => void): any;
+  setUser(user?: Partial<T>): any;
+  reset(): Promise<never>;
+  redirect(name: string): any;
+  onRedirect(listener: () => void): void;
+  strategy(): string;
+  registerStrategy(strategyName: string, strategy: object): void;
+  setStrategy(strategyName: string): void;
+  setUserToken(token: string): Promise<void>;
+  getRefreshToken(strategyName: string): string;
```
Comment (JoaoPedroAS51) [inline types/index.d.ts:45]: Same here.  `setRefreshToken(strategyName, token)`, `getRefreshToken(strategyName)` and `syncRefreshToken(strategyName)` were moved to RefreshToken class and refactored. You can use them like this `$auth.refreshToken.set(token)`, `$auth.refreshToken.get()` and `$auth.refreshToken.sync()`.

File: types/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  setRefreshToken(strategyName: string, token?: string): string;
+  syncRefreshToken(strategyName: string): string;
```
Comment (JoaoPedroAS51) [inline types/index.d.ts:47]: There is a new method `refreshTokens()`. Can you add it?

File: examples/with-typescript/pages/secure.vue [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        </b-card>
+      </b-col>
+      <b-col md="4">
+        <b-card title="Scopes" class="mb-2">
+          User:
+          <b-badge>{{ $auth.hasScope('user') }}</b-badge>Test:
+          <b-badge>{{ $auth.hasScope('test') }}</b-badge>Admin:
+          <b-badge>{{ $auth.hasScope('admin') }}</b-badge>
+        </b-card>
+        <b-card title="token">{{ $auth.token || '-' }}</b-card>
```
Comment (JoaoPedroAS51) [inline examples/with-typescript/pages/secure.vue:17]: Since we added Token class, use `$auth.token` here will cause an error preventing this page from loading.

Can you change to `$auth.token.get() || '-' `?

File: types/index.d.ts [hunk 17]
```diff
@@ -0,0 +1,75 @@
+// Type definitions for @nuxtjs/auth 4.8
+// Project: https://auth.nuxtjs.org
+// TypeScript Version: 3.1
+
```
Comment (JoaoPedroAS51) [inline types/index.d.ts:4]: Can you add the new classes? There are `Token`, `RefreshToken`, `TokenStatus`, `RefreshController` and `RequestHandler`.

File: types/index.d.ts (hunk 13 above)
Comment (NickBolles) [reply types/index.d.ts:33]: lol. I guess that's what happens when a PR sits for a while :/ I'll work on this. Thanks for the comments!!

File: types/index.d.ts (hunk 17 above)
Comment (NickBolles) [reply types/index.d.ts:4]: do we really need RefreshController and RequestHandler? is the developer expected to interact with them or are they private to the module?

**Comment by NickBolles** [comment]
Alright I think I've made those updates besides RefreshController and RequestHandler.

Side note: am I supposed to mark the comments as resolved? or leave that for reviewers to resolve?

File: types/index.d.ts (hunk 17 above)
Comment (JoaoPedroAS51) [reply types/index.d.ts:5]: We will add custom scheme support in v5. So developers can use these classes in their own schemes.

**Comment by t-eckert** [comment]
As a user, looking forward to seeing this merged!

File: types/requestHandler.d.ts [hunk 18]
```diff
@@ -0,0 +1,8 @@
+export default class RequestHandler {
+    constructor(auth: any);
+    $auth: any;
```
Comment (JoaoPedroAS51) [inline types/requestHandler.d.ts:3]: `$auth` shouldn't be the Auth class?

**Comment by NickBolles** [comment]
closing in favor of #622

**Comment by pi0** [comment]
Hi @NickBolles Nice work. I may bring some parts from your works but it will only cover internal TS rewrite and auto-generated types.

Are you interested in helping with a new typescript section in docs and creating a repository for with-typescript example? We can then link it from docs :)

**Comment by NickBolles** [comment]
Sure! Where/when is the typescript re-writes happening? I was tempted to just rename everything to .ts and add a compile step instead of adding the types. 

I'd be interested in helping out or reviewing the typescript migration.

**Comment by JoaoPedroAS51** [comment]
@NickBolles Hi! Check #621

**Comment by NickBolles** [comment]
> @NickBolles Hi! Check #621 

Perfect. Thanks!
